File: 2043-simpleBankSystem/bank.c
#include <stdbool.h>
#include <stdlib.h>

typedef struct {
  long long *balance;
  int size;
} Bank;

Bank *bankCreate(long long *balance, int balanceSize) {
  Bank *bank = malloc(sizeof(Bank));
  bank->balance = balance;
  bank->size = balanceSize;
  return bank;
}

bool validateAccount(Bank *obj, int account) {
  return account >= 1 && account <= obj->size;
}

bool sufficientFunds(Bank *obj, int account, long long money) {
  return obj->balance[account - 1] >= money;
}

bool bankTransfer(Bank *obj, int account1, int account2, long long money) {
  if (!validateAccount(obj, account1) || !validateAccount(obj, account2)) {
    return false;
  }
  if (!sufficientFunds(obj, account1, money)) {
    return false;
  }
  obj->balance[account1 - 1] -= money;
  obj->balance[account2 - 1] += money;
  return true;
}

bool bankDeposit(Bank *obj, int account, long long money) {
  if (!validateAccount(obj, account)) {
    return false;
  }
  obj->balance[account - 1] += money;
  return true;
}

bool bankWithdraw(Bank *obj, int account, long long money) {
  if (!validateAccount(obj, account)) {
    return false;
  }
  if (!sufficientFunds(obj, account, money)) {
    return false;
  }
  obj->balance[account - 1] -= money;
  return true;
}

void bankFree(Bank *obj) {}

/**
 * Your Bank struct will be instantiated and called as such:
 * Bank* obj = bankCreate(balance, balanceSize);
 * bool param_1 = bankTransfer(obj, account1, account2, money);

 * bool param_2 = bankDeposit(obj, account, money);

 * bool param_3 = bankWithdraw(obj, account, money);

 * bankFree(obj);
*/

File: 2043-simpleBankSystem/bank.hs
newtype Bank = Bank [Integer]

bankCreate :: [Integer] -> Bank
bankCreate = Bank

validateAccount :: Bank -> Int -> Bool
validateAccount (Bank accounts) accNum =
  accNum >= 1 && accNum <= length accounts

sufficientFunds :: Bank -> Int -> Integer -> Bool
sufficientFunds (Bank accounts) accNum = (>=) $ accounts !! accNum - 1

addTo :: [Integer] -> Int -> Integer -> [Integer]
addTo accounts accNum amount =
  let (before, x : after) = splitAt (accNum - 1) accounts
   in before ++ (x + amount) : after

withdrawFrom :: [Integer] -> Int -> Integer -> [Integer]
withdrawFrom accounts accNum amount = addTo accounts accNum (-amount)

bankTransfer :: Bank -> Int -> Int -> Integer -> (Bool, Bank)
bankTransfer (Bank accounts) acc1 acc2 amount
  | not (validateAccount (Bank accounts) acc1) || not (validateAccount (Bank accounts) acc2) = (False, Bank accounts)
  | not (sufficientFunds (Bank accounts) acc1 amount) = (False, Bank accounts)
  | otherwise =
      let updatedAccounts = do
            let afterWithdraw = withdrawFrom accounts acc1 amount
            addTo afterWithdraw acc2 amount
       in (True, Bank updatedAccounts)

bankDeposit :: Bank -> Int -> Integer -> (Bool, Bank)
bankDeposit (Bank accounts) accNum amount
  | not (validateAccount (Bank accounts) accNum) = (False, Bank accounts)
  | otherwise =
      let updatedAccounts = addTo accounts accNum amount
       in (True, Bank updatedAccounts)

bankWithdraw :: Bank -> Int -> Integer -> (Bool, Bank)
bankWithdraw (Bank accounts) accNum amount
  | not (validateAccount (Bank accounts) accNum) = (False, Bank accounts)
  | not (sufficientFunds (Bank accounts) accNum amount) = (False, Bank accounts)
  | otherwise =
      let updatedAccounts = withdrawFrom accounts accNum amount
       in (True, Bank updatedAccounts)
